src/dispatch_package/dispatch.cpp

```diff
@@ -0,0 +1,352 @@
+      + "bus_busy_start="     +std::to_string(bus_busy_start)    +", "
+      + "bus_busy_end="       +std::to_string(bus_busy_end)      +", "
+      + "bus_id="             +std::to_string(bus_id)            +", "
+      + "start_depot_id="     +std::to_string(start_depot_id)    +", "
+      + "end_depot_id="       +std::to_string(end_depot_id)      +", "
+      + "energy_left="        +std::to_string(energy_left)
+      + ">";
+  }
+};
+typedef std::vector<TripInfo> trips_t;
+
+trips_t csv_trips_to_internal(const std::string &inpstr){
+  trips_t trips;
+
+  std::stringstream ss;
+  ss<<inpstr;
+
+  io::CSVReader<7> in("trips.csv", ss);
+  in.read_header(io::ignore_extra_column, "trip_id","block_id","start_arrival_time","start_stop_id","end_arrival_time","end_stop_id","distance");
+
+  TripInfo temp;
+  while(in.read_row(temp.trip_id,temp.block_id,temp.start_arrival_time,temp.start_stop_id,temp.end_arrival_time,temp.end_stop_id,temp.distance)){
+    trips.emplace_back(temp);
+  }
+
+  return trips;
+}
+
+
+
+class Model {
+ private:
+  Parameters params;
+  const trips_t trips_template;
+  const stops_t stops;
+
+  static trips_t load_and_sort_trips(const std::string &trips_csv){
+    trips_t trips = csv_trips_to_internal(trips_csv);
+
+    //Sort trips into blocks where each block is ordered by start arrival time
+    std::stable_sort(trips.begin(), trips.end(), [](const auto &a, const auto &b){ return a.start_arrival_time<b.start_arrival_time; });
+    std::stable_sort(trips.begin(), trips.end(), [](const auto &a, const auto &b){ return a.block_id<b.block_id; });
+
+    return trips;
+  }
+
+ public:
+  Model(
+    const Parameters &params,
+    const std::string &trips_csv,
+    const std::string &stops_csv
+  ) : params(params),
+      trips_template(load_and_sort_trips(trips_csv)),
+      stops(csv_stops_to_internal(stops_csv))
+  {}
+
+  void update_params(const Parameters &new_params){
+    params = new_params;
+  }
+
+  ///Simulate bus movement along a route. Information about the journey is
+  ///stored by modifying the block.
+  ///
+  ///@param block_start Iterator pointing to the start of the block
+  ///@param block_end   Iterator pointing one past the end of the block
+  void run_block(trips_t::iterator &block_start, trips_t::iterator &block_end) const {
+    //Get time, distance, and energy from depot to route
+    const auto start_depot = stops.at(block_start->start_stop_id);
+    const auto energy_depot_to_start = start_depot.depot_distance*params.kwh_per_km;
+
+    block_start->start_depot_id = start_depot.depot_id;
+    block_start->energy_left    = params.battery_cap_kwh - energy_depot_to_start;
+    block_start->bus_id         = 1;
+
+    //Run through all the trips in the block
+    auto prevtrip=block_start;
+    for(auto trip=block_start;trip!=block_end;trip++){
+      //TODO: Incorporate charging at the beginning of trip if there is a charger present
+      auto trip_energy = trip->distance * params.kwh_per_km;
```

Comment:
perhaps we should log the full energy cycle of a block to evaluate for patterns that get us in the energy trap?

Reply:
Let's go over this logic with a fine-toothed comb shortly.